### PR TITLE
Fix data-toggle="buttons" when extra <input type="hidden">s are present

### LINKED ISSUE #16693: data-toggle="buttons" problem in v3.3.5
I don't know I think this issue is specific for Rails.

But bootstrap version up have broken `button toggle's function`.

I have a code like below.

```
<div data-toggle="buttons">
      <div class="checkbox">
        <label class="btn btn-block btn-facebook " for="post_post_facebook">
          <input name="post[post_facebook]" type="hidden" value="0" />
          <input class="hidden" type="checkbox" value="1" name="post[post_facebook]" id="post_post_facebook" /> Share with Facebook</label>
      </div>
</div>
```
[Gyazo](http://gyazo.com/e6692c1a3ff463b6f4237f690a8b79af)

The code works with version 3.3.4.1.

But, With 3.3.5. After click button. Toggle function does not works.

So, I have removed `<input name="post[post_facebook]" type="hidden" value="0" />` .

Then Toggle function works.

I think `data-toggle='buttons'` does not work with Rails form generator.

Is not?

（moved from https://github.com/twbs/bootstrap-sass/issues/923#issuecomment-114588190)

### PR DESCRIPTION
**Pull request by dsusco**
The check_box form helper generates a hidden input which was preventing the change event from being triggered. Specifying the type of the input to one of the two types we're looking for fixes the issue.

Fixes #16693. [EDIT: Added bug reference – Ed.]

### CODE REVIEW

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 5994e8bd2a9bba78ccff01d98d411b288d74cf45
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/68801807

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by patrickhlauke** [comment]
In principle, this change seems good to me (and I don't think it needs a separate/new test). @cvrebert / @fat ?

For future reference, you don't need to submit changes for the `dist/` or the minified version of the script.

**Comment by dsusco** [comment]
Thanks, I wasn't sure on the generated stuff so through it in just in case it was needed.

**Comment by cvrebert** [comment]
I always prefer adding a unit test.

**Comment by dsusco** [comment]
Added.

**Comment by cvrebert** [comment]
@dsusco Much appreciated!

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: ea9c9fd117c7ead47184fde684b80f8dccf8f506
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/68869886

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by cvrebert** [comment]
Need to check whether this is only applicable to invalid HTML like #16693.

**Comment by cvrebert** [comment]
I double-checked the HTML5 spec and having multiple `<input>`s inside one `<label>` like that indeed isn't valid HTML (as noted in #16693). So I'm mildly opposed to trying to accommodate bad markup like this.
Assuming you're using `rails-bootstrap-forms`, someone's already filed a bug about the bad markup: https://github.com/bootstrap-ruby/rails-bootstrap-forms/issues/231

CC: @fat @hnrch02 @XhmikosR @patrickhlauke for more opinions

**Comment by EHLOVader** [comment]
It might also be worth noting that setting the `checked` attribute of inputs of a type other than a `radio` or `checkbox` is also invalid. 

Which is what this PR aims to fix.  Might be worth thinking about then?

**Comment by dsusco** [comment]
The issue would still persist if the labels were another element (say a div). That would be valid HTML, if not overly semantic.

**Comment by cvrebert** [comment]
@dsusco Sure, but that's a pure hypothetical. I'm pretty sure the way the widget is coded requires `<label>`s rather `<div>`s.

@EHLOVader I don't quite see what your point is. There are a million ways to get any piece of markup wrong such that code manipulating its DOM will crash or behave improperly. The line between what to tolerate and what to not tolerate must be drawn somewhere; HTML5 validity would be one reasonable place to draw the line.

**Comment by claar** [comment]
@cvrebert Sure, drawing the line at validity is reasonable in general.

But this instance may warrant an exception, since outputting a hidden input is an extremely common workaround for getting browsers to POST unchecked checkboxes, as evidenced by Rails, Laravel (with Former), and I'm sure other frameworks having the same behavior.

In Laravel's Former (of which I'm the maintainer) for instance, the hidden input is output automatically when you output a checkbox, making this issue difficult to workaround for the developer.

**Comment by EHLOVader** [comment]
@cvrebert It is completely valid to put `input`, `button`, `select`,`textarea`, or `keygen` into a `label`.  If those are in a label processed by this, your DOM manipulation will blindly add `checked` property to the element. Which is invalid in all but the smallest case, `input`s with type of `checkbox` or `radio`. 

Adding this would prevent that and as a result preserve validity of HTML which is valid.

**Comment by dsusco** [comment]
@cvrebert I don't think it requires labels: https://jsfiddle.net/5kxkLa8g/

**Comment by Garbee** [comment]
@claar As far as doing it so it isn't POSTed. It doesn't matter whether it is within the label or just before it. It would be *best* to put the hidden just before the label that way the markup is valid. Forms are hard enough as it is, best not make them even more difficult to debug.

Looking over the PR, the actual change looks perfectly valid. Targeting only the types of inputs that the attribute can apply to as @EHLOVader pointed out. However, the unit test added is pure junk since it uses invalid markup which by default may cause problems. 

The tests should instead consist of more meaningful targets. To verify that checked isn't being applied to a `text` or `date` type for example.

**Comment by patrickhlauke** [comment]
FWIW i agree with @Garbee - the change to button.js itself is good in my opinion, as it explicitly checks that we're only doing things to the two types of inputs that can actually handle our code. And yes, the test, should then just verify that having other types of inputs will not cause erroneous additions of attributes.

**Comment by cvrebert** [comment]
@dsusco Hmm. That's not accessibility-friendly though.

@EHLOVader But why would you put a non-checkbox inside a Bootstrap checkbox widget in the first place? What's that even supposed to mean? Unless it's at least *alongside* an `<input type="checkbox">`...

**Comment by XhmikosR** [comment]
I think we should just stick to the specs here. Sure, people might use `div`s for POST stuff but IIRC it can be done with the `div` outside of the `label` too.

So, this PR seems fine to me too.

**Comment by NogsMPLS** [comment]
PR makes selection more explicit so that is good.

Are we not concerned that ```$.find()``` returns an array, and thus could be more than 1 input, and enabling invalid HTML to get through? Though, I have no idea why anyone would actually do that.

Coming at it from another perspective, perhaps an $input.length check just to make sure an actual valid input  exists?

**Comment by cvrebert** [comment]
Yeah, throwing an error if `$input.length != 1` is a good idea (see also #16324).

**Comment by cvrebert** [comment]
Hmm, apparently the validator (https://github.com/validator/validator/issues/112) and myself are on crack; [the spec exempts `type="hidden"`:](https://html.spec.whatwg.org/multipage/forms.html#category-label)
> **labelable elements**. These are elements that can be associated with a label element.
> => [...] `input` (**if the `type` attribute is not in the `hidden` state**), [...]

So I withdraw my objection. Sorry for the confusion!

File: dist/js/bootstrap.js [hunk 1]
```diff
@@ -227,7 +227,7 @@ if (typeof jQuery === 'undefined') {
     var $parent = this.$element.closest('[data-toggle="buttons"]')
 
     if ($parent.length) {
-      var $input = this.$element.find('input')
+      var $input = this.$element.find('input[type="radio"], input[type="checkbox"]')
```
Comment (cvrebert) [inline dist/js/bootstrap.js]: Could you add something like:
```js
if ($input.length != 1) {
  throw new Error('Checkbox/radio button group item must contain exactly 1 checkbox or radio input!')
}
```
as suggested by @NogsMPLS?

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: bc154d04ec0b1ca63550f4cf39f95385d36871a9
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/70103310

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by EHLOVader** [comment]
Was this updated for the new testing requirements?

**Comment by dsusco** [comment]
Not sure what you're referring to. I added the exception that NogsMPLS/cvrebert suggested.

**Comment by EHLOVader** [comment]
Absolutely, I saw that, and thank you @dsusco. I was trying to ask if that was enough for this to be accepted but I believe the right words didn't come out. 

Looks like @cvrebert said yes though by adding the milestone. :joy_cat:  

Thank you so much, all of you! for the work on this fix.

**Comment by EHLOVader** [comment]
I am still having issues with this. Since I use [SASS](https://github.com/twbs/bootstrap-sass) it isn't just as simple using this PR branch in my bower dependencies. 

I was reminded of this because I failed to find and replace the minified copy in my deployment recently and am just now hearing from users that the checkbox wasn't working. I also see this has gotten conflicts while waiting for approval. Any news on this?

**Comment by mdo** [comment]
[**Bootstrap 3 is no longer being officially developed or supported.**](https://github.com/twbs/bootstrap/issues/20631)

All work has moved onto our next major release, v4. As such, this issue or pull request is being closed as a "won't fix." For additional help and support, we recommend [utilizing our community resources](https://github.com/twbs/bootstrap/blob/master/README.md#community). Thanks for your understanding, and see you on the other side of v4!

<3,
@mdo and team

**Pull request closed without merging**